Unify error codes returned by various TLS backends

sweatybridge opened the pull request:
## Reconciliation of host name validation error
The [original issue](https://github.com/curl/curl/issues/2729) regarding inconsistent error code returned by different TLS backends during server certificate verification is caused by mishandling of framework return codes on host name validation failure. For openssl, host name is verified by a custom routine `verifyhost` which returns CURLE_PEER_FAILED_VERIFICATION on failure. For darwinssl, host name is verified automatically by the framework when calling `SSLHandle`. However, the framework does not return errSSLHostNameMismatch on failure as we might expect. Instead it returns a generic errSSLXCertChainInvalid which is mapped to CURLE_SSL_CACERT by curl. **So the first thing I would like to clarify is the intended difference between CURLE_PEER_FAILED_VERIFICATION and CURLE_SSL_CACERT, and whether we can use them interchangeably.**

The other inconsistency comes from curl's handling of schannel return code on Windows. Similar to darwinssl, the schannel framework also verifies host name automatically and returns SEC_E_WRONG_PRINCIPAL on validation failure. Although curl correctly catches this error code returned by `InitializeSecurityContext`, it only prints a different error message instead of mapping it to a return code consistent with darwinssl. We should fix this problem by returning CURLE_SSL_CACERT instead (commit 2e57a07).

## Reconciliation of other discrepancies
As we looked deeper into the implementation of schannel, darwinssl, and openssl, we also discovered several other sources of discrepancy. The details of what each discrepancy is and how we think it should be resolved is elaborated in the commit messages of this PR. In general, **I think the principle we should follow is to return as specific an error code as possible that’s also semantically consistent with other TLS backends**. Since not everyone can be familiar with all backends, I would suggest that every TLS backend should try its best to align the error code with that of openssl. If it's not possible to be consistent, the difference should be documented. For example,
```
CURLE_SSL_ISSUER_ERROR (83)
	Issuer check failed. (Added in 7.19.0, only returned by curl when compiled with openssl backend).
```

## Test plan
https://badssl.com hosts an extensive set of URLs with intentionally misconfigured SSL certificates. We ran a test suite of curl compiled with different TLS backends against those URLs. The return codes from the different curl versions on different operating systems are consistent.

bagder commented:
> the intended difference between CURLE_PEER_FAILED_VERIFICATION and CURLE_SSL_CACERT

I'm not sure the distinction was ever really clear. The **libcurl-errors.3** man page certainly doesn't help much:

+ `CURLE_PEER_FAILED_VERIFICATION (51)`  - The remote server's SSL certificate or SSH md5 fingerprint was deemed not OK.
+ `CURLE_SSL_CACERT (60)` - Peer certificate cannot be authenticated with known CA certificates.

Given this, I would say that the latter is strictly for checks of the CA, so a mismatched hostname should perhaps use the former. But I also don't think splitting the errors like this is necessary or helps users. I think it is rather detrimental and we should try to use one single error for "not allowed to connect due to failed verification".

sweatybridge commented on lib/vtls/darwinssl.c:
```diff
@@ -2195,30 +2186,32 @@ static int verify_cert(const char *cafile, struct Curl_easy *data,
   }
   else if(ret != noErr) {
     CFRelease(array);
-    return sslerr_to_curlerr(data, ret);
```
In other places where `SSLCopyPeerTrust` is called, curl simply ignores the error code returned. But I agree with you that printing an error message would help with diagnosis when there's an failure so I'm brining back `failf`.

sweatybridge commented on lib/vtls/darwinssl.c:
```diff
@@ -2389,28 +2427,49 @@ darwinssl_connect_step2(struct connectdata *conn, int sockindex)
+      case errSSLClientCertRequested:
+        failf(data, "The server has requested a client certificate.");
+        /* This code should not be returned because
+           kSSLSessionOptionBreakOnCertRequested is never set */
```
I don't know if the implementation might change in the future (either curl starts setting the option or SecureTransport decides to remove the option to always break when server requests cert). Although these circumstances are very unlikely, we might want to log the error just in case.

Alternatively I can change it to a comment if you prefer.

sweatybridge commented:
@nickzman I've modified the relevant commits based on your suggestions. The error message style looks consistent now. I've also checked the available enums against https://github.com/phracker/MacOSX-SDKs/blob/master/MacOSX10.5.sdk/System/Library/Frameworks/Security.framework/Versions/A/Headers/SecureTransport.h#L164 to make sure that newer additions are guarded by the appropriate macro.

@bagder I think your explanation makes a lot of sense. I've changed the return code for SEC_E_WRONG_PRINCIPAL to CURLE_PEER_FAILED_VERIFICATION in the last commit since it represents a more general problem than the CA certificate (and more specific than the original CURLE_SSL_CONNECT_ERROR). I think such a design is acceptable as long as it's documented that one code is strictly a subproblem of the other (i.e. CURLE_SSL_CACERT < CURLE_PEER_FAILED_VERIFICATION < CURLE_SSL_CONNECT_ERROR). Sometimes it might even be useful. For example, users of libcurl may want to programmatically handle very specific error cases based on the code returned.

If you agree, I can change the return code for errSSLXCertChainInvalid (darwinssl) and [SSL_R_CERTIFICATE_VERIFY_FAILED](https://github.com/sweatybridge/curl/blob/ba58ce669ca6609253055d081a00a4f709ce382e/lib/vtls/openssl.c#L2712) (openssl) to CURLE_PEER_FAILED_VERIFICATION as well for consistency. But we should beware that such a fix will likely affect many users because those error codes are fairly common. Do you have a plan on how we can manage the risk of this change?

bagder commented:
Since openssl is the by far most used SSL backend I think we need to view that's behavior as the default one and the one we should try to align the other backends to. That way we fix more problems than we introduce.

Thinking about these return codes more, I'm pretty much against changing anything with openssl that *was* `CURLE_SSL_CACERT` to something else, since I'm pretty sure that's the specific return code more than one application checks for to detect CA problems as that's what the OpenSSL backend has returned for those. Even the curl command line tool does this!

But I find the error code name to be a bit misleading. Maybe we should drop the distinctions and just have a single return code for "failed peer verification"? Something along these lines (although this change alone causes compiler warnings) ?

~~~diff
diff --git a/include/curl/curl.h b/include/curl/curl.h
index 067b34ded..4cd55cc89 100644
--- a/include/curl/curl.h
+++ b/include/curl/curl.h
@@ -515,22 +515,21 @@ typedef enum {
   CURLE_OBSOLETE46,              /* 46 - NOT USED */
   CURLE_TOO_MANY_REDIRECTS,      /* 47 - catch endless re-direct loops */
   CURLE_UNKNOWN_OPTION,          /* 48 - User specified an unknown option */
   CURLE_TELNET_OPTION_SYNTAX,    /* 49 - Malformed telnet option */
   CURLE_OBSOLETE50,              /* 50 - NOT USED */
-  CURLE_PEER_FAILED_VERIFICATION, /* 51 - peer's certificate or fingerprint
-                                     wasn't verified fine */
+  CURLE_OBSOLETE51,              /* 51 - NOT USED from 7.62.0 */
   CURLE_GOT_NOTHING,             /* 52 - when this is a specific error */
   CURLE_SSL_ENGINE_NOTFOUND,     /* 53 - SSL crypto engine not found */
   CURLE_SSL_ENGINE_SETFAILED,    /* 54 - can not set SSL crypto engine as
                                     default */
   CURLE_SEND_ERROR,              /* 55 - failed sending network data */
   CURLE_RECV_ERROR,              /* 56 - failure in receiving network data */
   CURLE_OBSOLETE57,              /* 57 - NOT IN USE */
   CURLE_SSL_CERTPROBLEM,         /* 58 - problem with the local certificate */
   CURLE_SSL_CIPHER,              /* 59 - couldn't use specified cipher */
-  CURLE_SSL_CACERT,              /* 60 - problem with the CA cert (path?) */
+  CURLE_PEER_FAILED_VERIFICATION, /* 60 - not verified okay */
   CURLE_BAD_CONTENT_ENCODING,    /* 61 - Unrecognized/bad encoding */
   CURLE_LDAP_INVALID_URL,        /* 62 - Invalid LDAP URL */
   CURLE_FILESIZE_EXCEEDED,       /* 63 - Maximum file size exceeded */
   CURLE_USE_SSL_FAILED,          /* 64 - Requested FTP SSL level failed */
   CURLE_SEND_FAIL_REWIND,        /* 65 - Sending the data requires a rewind
@@ -582,10 +581,13 @@ typedef enum {
   CURLE_RECURSIVE_API_CALL,      /* 93 - an api function was called from
                                     inside a callback */
   CURL_LAST /* never use! */
 } CURLcode;
 
+/* added in 7.62.0 */
+#define CURLE_SSL_CACERT CURLE_PEER_FAILED_VERIFICATION
+
 #ifndef CURL_NO_OLDIES /* define this to test if your app builds with all
                           the obsolete stuff removed! */
 
 /* Previously obsolete error code re-used in 7.38.0 */
 #define CURLE_OBSOLETE16 CURLE_HTTP2
~~~

sweatybridge commented:
> the specific return code more than one application checks for to detect CA problems

Indeed, that was my concern as well. Just beware that changing the error code the other way like you suggested poses the same problem but to a lesser extent since CURLE_PEER_FAILED_VERIFICATION was introduced later and should have less users depending on it.

> just have a single return code for "failed peer verification"

That would be the best scenario. I've made the necessary changes to get it to compile. Hopefully it won't break any users. If we want to be safe, I can also move that commit to another PR so that it can be easily reverted if things go south. Otherwise, I think this PR is ready @nickzman.

nickzman commented on lib/vtls/darwinssl.c:
```diff
@@ -2379,6 +2377,53 @@ darwinssl_connect_step2(struct connectdata *conn, int sockindex)
+      case errSSLCrypto:
+        failf(data, "An underlying cryptographic error was encountered");
+        break;
+#if CURL_BUILD_MAC_10_11
```
errSSLWeakPeerEphemeralDHKey is available in the iOS 9 & later SDK, so let's open it up to iOS users as well please.

nickzman commented on lib/vtls/darwinssl.c:
```diff
@@ -2379,6 +2377,53 @@ darwinssl_connect_step2(struct connectdata *conn, int sockindex)
+        break;
+#if CURL_BUILD_MAC_10_11
+      case errSSLWeakPeerEphemeralDHKey:
+        failf(data, "Indicates a weak ephemeral dh key");
```
That should probably be "Diffie-Hellman" or "D-H" and not "dh."

nickzman commented on lib/vtls/darwinssl.c:
```diff
@@ -2419,30 +2480,98 @@ darwinssl_connect_step2(struct connectdata *conn, int sockindex)
+      case errSSLPeerUnexpectedMsg:
+        failf(data, "Peer rejected unexpected message");
+        break;
+#if CURL_BUILD_MAC_10_11
```
This one's also available in the iOS 9 SDK.

sweatybridge commented:
PR updated but build failed due to flaky test. Can someone trigger the build again? It passes on my personal CI https://travis-ci.org/sweatybridge/curl

bagder commented:
Thanks!

I'm fine with the changes, but due to the nature of the changes I prefer to wait until after the release before I merge this PR.

bagder commented:
Thanks!

mback2k commented:
Unfortunately this change breaks lagacy MinGW / Windows XP compatibility, please check the following error message:
```
vtls/schannel.c: In function 'schannel_connect_step1':
vtls/schannel.c:813:12: error: 'SEC_E_APPLICATION_PROTOCOL_MISMATCH' undeclared (first use in this function)
       case SEC_E_APPLICATION_PROTOCOL_MISMATCH:
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
vtls/schannel.c:813:12: note: each undeclared identifier is reported only once for each function it appears in
vtls/schannel.c: In function 'schannel_connect_step2':
vtls/schannel.c:1050:14: error: 'SEC_E_APPLICATION_PROTOCOL_MISMATCH' undeclared (first use in this function)
         case SEC_E_APPLICATION_PROTOCOL_MISMATCH:
              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

mback2k commented on lib/vtls/schannel.c:
```diff
@@ -669,14 +673,20 @@ schannel_connect_step1(struct connectdata *conn, int sockindex)
 
     if(sspi_status != SEC_E_OK) {
-      if(sspi_status == SEC_E_WRONG_PRINCIPAL)
-        failf(data, "schannel: SNI or certificate check failed: %s",
```
Why was this conditional error message removed here?

bagder replied:
See also #2950

bagder replied:
Are you saying that failf() was useful there and we should reinstate it?

sweatybridge replied:
As mentioned in the commit message, `AcquireCredentialsHandle` doesn't return SEC_E_WRONG_PRINCIPAL https://msdn.microsoft.com/en-us/library/windows/desktop/aa374716(v=vs.85).aspx. So I think the previous error message was dead code, unless there is a legacy use case that I overlooked...

lietusme commented:
As I understand looking briefly - CURLE_SSL_CACERT is deprecated and should no longer be used?

It broke our switch statement because CURLE_SSL_CACERT is same as CURLE_PEER_FAILED_VERIFICATION now. 

If its deprecated, it should be documented better, as it now says this:

```
/* added in 7.62.0 */
#define CURLE_SSL_CACERT CURLE_PEER_FAILED_VERIFICATION
```

sweatybridge commented:
Yes, more documentation is better. I can move it to the `CURL_NO_OLDIES` ifndef and add a bit more context to the comment. Someone might also want to update the documentation page here: https://curl.haxx.se/libcurl/c/libcurl-errors.html

sweatybridge commented:
#3291
